zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -464,1380 +464,231 @@ pub(crate) fn put_received_note<T: ReceivedSaplingOutput>(
-            .build();
-
-        let (account, usk, _) = st.test_account().unwrap();
-        let dfvk = st.test_account_sapling().unwrap();
-
-        // Add funds to the wallet in a single note
-        let value = NonNegativeAmount::const_from_u64(50000);
-        let (h1, _, _) = st.generate_next_block(&dfvk, AddressType::DefaultExternal, value);
-        st.scan_cached_blocks(h1, 1);
-
-        // Spendable balance matches total balance at 1 confirmation.
-        assert_eq!(st.get_total_balance(account), value);
-        assert_eq!(st.get_spendable_balance(account, 1), value);
-
-        let extsk2 = ExtendedSpendingKey::master(&[]);
-        let addr2 = extsk2.default_address().1;
-        let to = addr2.into();
-
-        // TODO: This test was originally written to use the pre-zip-313 fee rule
-        // and has not yet been updated.
-        #[allow(deprecated)]
-        let fee_rule = StandardFeeRule::PreZip313;
-
-        #[allow(clippy::type_complexity)]
-        let send_and_recover_with_policy = |st: &mut TestState<BlockCache>,
-                                            ovk_policy|
-         -> Result<
-            Option<(sapling::Note, PaymentAddress, MemoBytes)>,
-            Error<
-                SqliteClientError,
-                commitment_tree::Error,
-                GreedyInputSelectorError<Zip317FeeError, ReceivedNoteId>,
-                Zip317FeeError,
-            >,
-        > {
-            let min_confirmations = NonZeroU32::new(1).unwrap();
-            let proposal = st.propose_standard_transfer(
-                account,
-                fee_rule,
-                min_confirmations,
-                &to,
-                NonNegativeAmount::const_from_u64(15000),
-                None,
-                None,
-                ShieldedProtocol::Sapling,
-            )?;
-
-            // Executing the proposal should succeed
-            let txid = st.create_proposed_transactions(&usk, ovk_policy, &proposal)?[0];
-
-            // Fetch the transaction from the database
-            let raw_tx: Vec<_> = st
-                .wallet()
-                .conn
-                .query_row(
-                    "SELECT raw FROM transactions
-                    WHERE txid = ?",
-                    [txid.as_ref()],
-                    |row| row.get(0),
-                )
-                .unwrap();
-            let tx = Transaction::read(&raw_tx[..], BranchId::Canopy).unwrap();
+        fn with_decrypted_pool_memos(
+            d_tx: &DecryptedTransaction<'_, AccountId>,
+            mut f: impl FnMut(&MemoBytes),
+        ) {
+            for output in d_tx.sapling_outputs() {
+                f(output.memo());
+            }
+        }
 
+        fn try_output_recovery<Cache>(
+            st: &TestState<Cache>,
+            height: BlockHeight,
+            tx: &Transaction,
+            fvk: &Self::Fvk,
+        ) -> Result<Option<(Note, Address, MemoBytes)>, OutputRecoveryError> {
```

Comment:
nit: it might be possible to just provide a default implementation for this, if we instead make a method to construct the `Domain`? We don't have an `orchard` equivalent for `try_sapling_output_recovery` so we might just do this when creating the Orchard impl.

Reply:
Yep, we might find a nicer way to generalise this, but for now this is fine.